File: online_store/domain/models/products/Electronic.py
from online_store.domain.models.products.ProductInterface import ProductInterface

class Electronic(ProductInterface):
    def __init__(self, name: str, brand: str, price: float):
        self.name = name
        self.brand = brand
        self.price = price

    def getPrice(self) -> float:
        return self.price
    
    def modifyPrice(self, new_price:float):
        self.price = new_price

    def getTitle(self) -> str:
        return self.name
    
    def clone(self):
        return Electronic(self.name, self.brand, self.price)
    
    def __str__(self):
        return f"{self.name} by {self.brand} - Price: ${self.price:.2f}"

File: tmpsjava/online_shop/src/domain/models/products/Electronic.java
package domain.models.products;

import java.util.ArrayList;
import java.util.List;

public class Electronic implements ProductInterface{
    final private String title;
    private float price;

    private Promotion promotion;

    private List<Observer> observers;

    public Electronic(String title, float price, Promotion promotion){
        this.title = title;
        this.price = price;
        this.promotion = promotion;
        this.observers = new ArrayList<>();
    }

    @Override
    public float getPrice() {
        return this.promotion.applyDiscount(this.price);
    }

    @Override
    public void modifyPrice(float new_price) {
        this.price = new_price;
        notifyObservers("Price changed: $" + new_price);
    }

    @Override
    public String getTitle() {
        return this.title;
    }

    @Override
    public ProductInterface copy() {
        return (ProductInterface) new Book(this.title, this.price, this.promotion);
    }

    @Override
    public Promotion getPromotion() {
        return promotion;
    }

    @Override
    public void addObserver(Observer observer) {
        observers.add(observer);
    }

    @Override
    public void removeObserver(Observer observer) {
        observers.remove(observer);
    }

    @Override
    public void notifyObservers(String message) {
        for (Observer observer : observers) {
            observer.update(this.title, message);
        }
    }
}
